### PR TITLE
Bump to share

### PR DESCRIPTION
**Pull request by pkendall64**
This PR is an extension to the "Loan Model" functionality for transmitter modules with an accelerometer.

The person who is loaning their model must be connected to the model that wish to loan out.

With your radios in an upright position (i.e. antenna facing the sky) bump the radios together. The lender's module will send the "Loan Model" command to the RX and the borrower's TX module will start binding. If all goes well, the model is loaned to the borrower.

To return the model, the borrower uses the LUA script to "Return Model" or the owner power cycles 3 times, as per the "Loan Model" functionality.

### CODE REVIEW

**Comment by CapnBry** [comment]
I'm trying to test this and I am smacking the heck out of my radios (Axis Thor on one) with the antennas pointing up w/ a RX connected and it never goes into binding mode. Is there some other criteria I am missing? Using "Loan Model" from the Lua works, but I can't seem to get a proper bump out of it. Maybe I don't know what a bump is? To me it means moving the backs (module side) toward each other until they hit each other. I've done it a lot harder than someone with a $140 ExpressLRS module probably ever should.

Oh snap they're on sale $95 now! (smacks the heck out of it recklessly)

**Comment by pkendall64** [comment]
> I'm trying to test this and I am smacking the heck out of my radios (Axis Thor on one) with the antennas pointing up w/ a RX connected and it never goes into binding mode. Is there some other criteria I am missing? Using "Loan Model" from the Lua works, but I can't seem to get a proper bump out of it. Maybe I don't know what a bump is? To me it means moving the backs (module side) toward each other until they hit each other. I've done it a lot harder than someone with a $140 ExpressLRS module probably ever should.
> 
> Oh snap they're on sale $95 now! (smacks the heck out of it recklessly)

I just knock it on the side with my knuckle!

**Comment by StonedDawg** [comment]
Tested this with axis bound to jumper RX, all is working as it should.... was only a tap with my finger!

File: src/lib/GSENSOR/devGsensor.cpp
Comment (CapnBry) [inline src/lib/GSENSOR/devGsensor.cpp:24]: unsigned

File: src/lib/GSENSOR/devGsensor.cpp
Comment (CapnBry) [inline src/lib/GSENSOR/devGsensor.cpp:74]: Should this somehow be protected to prevent borrowing my own loan? 2 bumps 400ms apart mean I both loan and borrow my own model. 
```

UART STATS Bad:Good = 0:500
UART STATS Bad:Good = 0:500
UART STATS Bad:Good = 0:500
Loaning model
Bumps 1 : 0.183 -1.87 0.193  // <-- my bump

UART STATS Bad:Good = 0:500
Borrowing model
Bumps 1 : 0.33 0.974 0.3 // <-- putting down the controller

UART STATS Bad:Good = 0:500
UART STATS Bad:Good = 0:500
hwTimer stop
set rate 5
hwTimer interval: 20000
Config LoRa
Adjusted max packet size 64-64
hwTimer resume
Entered binding mode at freq = 12302619
$X<     /¤Îw!set rate 2
hwTimer interval: 2000
Config LoRa
Adjusted max packet size 32-32
Exiting binding mode
UART STATS Bad:Good = 0:428
UART STATS Bad:Good = 0:455
UART STATS Bad:Good = 0:500
```

**Review by CapnBry — COMMENTED** [review]
I evaluated this again after manually repulling the PR and bumps work for me!

However, now when I put my radio down on the table (standing up because the massive Thor module means it can't lay down) it goes into loan mode, if I pick it up too quickly, it goes into loan mode, if I hit my right hand too hard with the controller as I prepare to use it or my finger taps the radio body, it goes into loan mode. Is there any way to _reduce_ the sensitivity?

This is a fun idea but I'm really concerned about it being a nuisance. It is really easy to activate for a function that will be so rarely used.

File: src/lib/GSENSOR/devGsensor.cpp
Comment (pkendall64) [reply src/lib/GSENSOR/devGsensor.cpp:58]: Ok, I'm with you on that one 😄

File: src/lib/GSENSOR/devGsensor.cpp
Comment (pkendall64) [reply src/lib/GSENSOR/devGsensor.cpp:74]: So what do you think about putting a stop to bump detection for 10s (or some other time period) after entering loan mode by bumping.

**Comment by CapnBry** [comment]
Any comments about the other points I've made? There's still some pending conversations in the "Files changed" view

File: src/src/tx_main.cpp
Comment (pkendall64) [reply src/src/tx_main.cpp:721]: I did think about that at the time, but the bind MSP only has 4 bytes of the UID and all zeros could be a real UID!

File: src/lib/GSENSOR/gsensor.cpp
Comment (pkendall64) [reply src/lib/GSENSOR/gsensor.cpp:102]: You are right I should remove that as it was when I was doing checks when the interrupt occurred which didn't work out.

**Review by CapnBry — APPROVED** [review]

**Comment by CapnBry** [comment]
This looks good, but do you want to maybe tie it to `config.getMotionMode() == 1`? I feel like there should be a way to disable it, and typing it to motionmode would make the most sense.

Also I noticed this (unrelated to your changes), not that it makes any difference but the fanmode and motionmode are not per-model settings so they should go outside the loop. Should we sneak that in too? If you make the bump tied to motion mode, I can claim it is related! 😁
image

**Comment by pkendall64** [comment]
> This looks good, but do you want to maybe tie it to `config.getMotionMode() == 1`? I feel like there should be a way to disable it, and typing it to motionmode would make the most sense.
> 
> ~Also I noticed this (unrelated to your changes), not that it makes any difference but the fanmode and motionmode are not per-model settings so they should go outside the loop. Should we sneak that in too? If you make the bump tied to motion mode, I can claim it is related!~ 😁 EDIT: I have this in my PWM update since I needed to change config anyway. image

Good ideas, implemented.
I guess we'll see who gets in first 😉

**Comment by CapnBry** [comment]
What's going on developers! Poor PK has been waiting for your approval for over a month! I still stand by my approval.

**Review by schugabe — CHANGES_REQUESTED** [review]
I did test this PR and the loaning starts even if arm switch is high. If you are flying with the antenna pointing up and tab the tx it would be a insta FS

**Comment by pkendall64** [comment]
> I did test this PR and the loaning starts even if arm switch is high. If you are flying with the antenna pointing up and tab the tx it would be a insta FS

What a disaster that would have been 😓

**Pull request merged**
